### DISCUSSION TITLE
Processing record before responding

### DISCUSSION THREAD

**Discussion by JunoNgx** [pocketbase/pocketbase#4668] [Q&A] [answered]
This is a somewhat duplication of #2790, but I didn't find the answer I needed.

I'm using JS hooks, if it matters.

What I'd like to happen: the user sends a url as a `POST` request. The backend Pocketbase instance will fetch the title and favicon link of this url and return everything in the response.

What I do: intercepting the `onRecordBeforeCreateRequest` hook. What goes in `funcs.processNewItem(e)` is a long process of passing data from one function to another.

```
onRecordBeforeCreateRequest(async (e) => {
    const funcs = require(`${__hooks}/funcs.js`);
    
    e.Record.set("type", "todo");
    funcs.processNewItem(e);
}, types.DbTables.ITEMS);
```

I also duly create a form and update the record with validation, if it matters:
```
const form = RecordUpsertForm($app, itemRecord);
form.loadData({
     title,
     content: processedUrl,
     faviconUrl: processedFaviconUrl
});
form.submit();
```

What I expect: `title`, `faviconUrl`, and `type` (as a `link`) fully filled and responded as part of the only `POST` request.

What happened:  pocketbase responds to the POST request exactly what I sent, which has only the field `content` filled, and all of `title`, `faviconUrl`, `type` are empty.

I am able to see the data I processed if I perform another `GET` request momentarily after, but never from the first `POST` request.

Am I missing something?

**Comment by ganigeorgiev** [comment]
I'm confused. How the 2 code snippets are related? Are you saving the record manually with the form in addition to the `onRecordBeforeCreateRequest`? Could you provide a more complete code sample to understand the broader picture?

**Reply by JunoNgx** [reply]
The second snippet is part of `funcs.processNewItem(e)`, very deep down the chain of passed functions.

Regardless, this is my `main.pb.js`: https://github.com/JunoNgx/justjot-backend/blob/master/pb_hooks/main.pb.js

And my `processNewItem`: https://github.com/JunoNgx/justjot-backend/blob/master/pb_hooks/funcs.js#L129

**Reply by ganigeorgiev** [reply, marked as answer]
You are creating a new Record instance in https://github.com/JunoNgx/justjot-backend/blob/master/pb_hooks/funcs.js#L131.

If you want to modify the Record instance of the request you can work directly with `e.model` or replace `e.model = itemRecord`.

But even then I would recommend to reconsider your execution flow because it seems that you are currently performing multiple db updates on the same object during the create hook execution which could have some side-effects - once with your manual RecordUpsert in `setItemAsTodo`/`setItemAsLink`, another one in `tryGetTitleAndFavicon` and again in the `onRecordBeforeCreateRequest` finalizer (the default).

**Reply by JunoNgx** [reply]
> If you want to modify the Record instance of the request you can work directly with e.model or replace e.model = itemRecord.

Was trying to follow your instruction here; took me a while to figure out what went wrong, but `onRecordBeforeCreateRequest` and `onRecordBeforeModel` have different data properties. For `onRecordBeforeCreateRequest`, I should access `e.record`.

> You are creating a new Record instance in https://github.com/JunoNgx/justjot-backend/blob/master/pb_hooks/funcs.js#L131.

This, however, is the main crux of the problem.

Would definitely consider the execution flow as well.

Thanks a lot for the assistance.
